### PR TITLE
#168781631: assign user roles

### PR DESCRIPTION
**Pull request by alainmateso**
#### What does this PR do?
Assign user roles
#### Description of Task to be completed?
- Admin can view all users
- Admin can view a specific user
- Admin can assign a specific user a role

#### How should this be manually tested?

- Clone this branch
- run npm i
- run npm run dev
- in your browser navigate to localhost:<:port>/admin/users
- select any user of your choice
- select a new role
- press assign new role button

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
#168781631
#### Screenshots (if appropriate)
Screen Shot 2019-12-18 at 16 14 16
Screen Shot 2019-12-18 at 16 14 53
Screen Shot 2019-12-18 at 16 15 02
Screen Shot 2019-12-18 at 16 17 06

#### Questions:
N/A

### CODE REVIEW

**Review by Cheza-Dzabala — COMMENTED** [review]
@alainmateso, great job! However, for the view that displays users to the admin, is it possible to use a data-table instead of the current way of displaying. You would have less code that does more. Maybe something like [this](https://www.npmjs.com/package/react-data-table-component).

File: src/components/pages/admin/ViewSpecificUser.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      [name]: value,
+    });
+  }
+
+  handleSubmit = async (e) => {
+    e.preventDefault();
+    const { Role } = this.state;
+    const { assignRole: assignNewRole, userId } = this.props;
+    await assignNewRole(userId, Role);
+    // window.location.reload();
```
Comment (sabin18) [inline src/components/pages/admin/ViewSpecificUser.js]: remove this line.

File: src/App.js [hunk 2]
```diff
@@ -42,10 +41,11 @@ export class App extends Component {
           <ProtectedRoute exact path="/requests/create" component={CreateRequest} />
           <ProtectedRoute exact path="/requests" component={ViewRequests} />
           <ProtectedRoute path="/requests/:requestId" component={SingleRequest} />
-          {/* <ProtectedRoute exact path="/accommodations" component={AllAccommodations} /> */}
           <ProtectedRoute path="/accommodations/new" component={NewAccommodation} />
           <ProtectedRoute exact path="/accommodations" component={GetAllAccommodations} />
           <ProtectedRoute exact path="/accommodations/:slug" component={GetsingleAccommodation} />
+          <ProtectedRoute exact path="/admin/users" component={ViewAllUsers} />
+          <ProtectedRoute exact path="/admin/users/:userId" component={SpecificUSer} />
```
Comment (nakiwuge) [inline src/App.js]: Use `username` instead of `userId` param

File: src/actions/userActions.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+export const getSpecificUser = (userId) => async (dispatch) => {
+  try {
+    const res = await backendCall.get(`/admin/users/${userId}`, { headers });
+    dispatch(userType(GET_SPECIFIC_USER, res.data));
+  } catch (error) {
+    dispatch(userType(GET_SPECIFIC_USER_ERROR, error.response));
+  }
+};
+
+export const getRoles = () => async (dispatch) => {
```
Comment (nakiwuge) [inline src/actions/userActions.js]: Move the roles actions to `roleAction` file

File: src/components/pages/admin/ViewSpecificUser.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    this.setState({
+      [name]: value,
+    });
+  }
+
+  handleSubmit = async (e) => {
+    e.preventDefault();
+    this.setState({ isLoadingButton: true });
+    const { Role } = this.state;
+    const { assignRole: assignNewRole, userId, getSpecificUser: getOneUser } = this.props;
```
Comment (nakiwuge) [inline src/components/pages/admin/ViewSpecificUser.js:48]: Move the constants above line 39

File: src/helpers/authHelper.js [hunk 5]
```diff
@@ -17,7 +17,19 @@ export const checkSupplier = () => {
   }
   return false;
 };
+export const checkAdmin = () => {
+  const userInfo = isAuth();
+  if (userInfo) {
+    const { payload } = userInfo;
+    const { role } = payload;
+    if (role !== 1) {
```
Comment (nakiwuge) [inline src/helpers/authHelper.js:25]: Add a new line

File: src/reducers/userReducer.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        userData: payload,
+        status: 'Success',
+      };
+    case GET_SPECIFIC_USER_ERROR:
+      return {
+        ...state,
+        userError: payload,
+        status: 'Failure',
+      };
+    case FETCH_ROLES:
```
Comment (nakiwuge) [inline src/reducers/userReducer.js]: Move the roles to a rolesReducer

File: src/views/admin/ViewAllUsers.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+/* eslint-disable react/prefer-stateless-function */
+import React from 'react';
+import { Redirect } from 'react-router-dom';
+import ViewUsers from '../../components/pages/admin/ViewUsers';
+import { checkAdmin } from '../../helpers/authHelper';
+
+
```
Comment (nakiwuge) [inline src/views/admin/ViewAllUsers.js]: remove this line

File: src/views/admin/ViewAllUsers.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+export const ViewAllUsers = () => (checkAdmin() ? (<ViewUsers />) : <Redirect to="/profile" />);
+
+
```
Comment (nakiwuge) [inline src/views/admin/ViewAllUsers.js]: remove this line

**Review by nakiwuge — CHANGES_REQUESTED** [review]
Good job @alainmateso , however I would like you to work on the feedback below.
- Remove the user ids from the cards where you are displaying users on `admin/users`
- Make the username on the card a heading so that it is bold
- Validate the `select role field` so that the user gets a descriptive error when he/she clicks `Assign role` without selecting a role. Currently, the error comes from the backend but it is too long
- Your coverage should be 90% plus
- check out for other comments  I have left on the pr

**Review by Cheza-Dzabala — APPROVED** [review]
Great job

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**